File: urbanairship-sdk/src/test/java/com/urbanairship/analytics/MediaEventTemplateTest.java
/* Copyright 2017 Urban Airship and Contributors */

package com.urbanairship.analytics;

import com.urbanairship.BaseTestCase;
import com.urbanairship.TestApplication;

import org.json.JSONException;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;

import static org.mockito.Mockito.mock;

public class MediaEventTemplateTest extends BaseTestCase {
    @Rule
    public ExpectedException exception = ExpectedException.none();

    Analytics analytics;

    @Before
    public void setup() {
        analytics = mock(Analytics.class);
        TestApplication.getApplication().setAnalytics(analytics);
    }

    /**
     * Test browsed event.
     *
     * @throws JSONException
     */
    @Test
    public void testBrowsedEventBasic() throws JSONException {
        CustomEvent event = MediaEventTemplate.newBrowsedTemplate().createEvent();

        EventTestUtils.validateEventValue(event, "event_name", MediaEventTemplate.BROWSED_CONTENT_EVENT);
        EventTestUtils.validateEventValue(event, "template_type", "media");
        EventTestUtils.validateNestedEventValue(event, "properties", "ltv", "false");
    }

    /**
     * Test browsed content event with optional properties.
     *
     * @throws JSONException
     */
    @Test
    public void testBrowsedEvent() throws JSONException {
        CustomEvent event = MediaEventTemplate.newBrowsedTemplate()
                                              .setCategory("media-category")
                                              .setId("starred-content-ID 1")
                                              .setDescription("This is a starred content media event.")
                                              .setType("audio type")
                                              .setAuthor("The Cool UA")
                                              .setFeature(true)
                                              .setPublishedDate("November 4, 2015")
                                              .createEvent();

        EventTestUtils.validateEventValue(event, "event_name", MediaEventTemplate.BROWSED_CONTENT_EVENT);
        EventTestUtils.validateEventValue(event, "template_type", "media");
        EventTestUtils.validateNestedEventValue(event, "properties", "ltv", "false");
        EventTestUtils.validateNestedEventValue(event, "properties", "category", "\"media-category\"");
        EventTestUtils.validateNestedEventValue(event, "properties", "id", "\"starred-content-ID 1\"");
        EventTestUtils.validateNestedEventValue(event, "properties", "description", "\"This is a starred content media event.\"");
        EventTestUtils.validateNestedEventValue(event, "properties", "type", "\"audio type\"");
        EventTestUtils.validateNestedEventValue(event, "properties", "author", "\"The Cool UA\"");
        EventTestUtils.validateNestedEventValue(event, "properties", "feature", "true");
        EventTestUtils.validateNestedEventValue(event, "properties", "published_date", "\"November 4, 2015\"");
    }

    /**
     * Test starred content event.
     *
     * @throws JSONException
     */
    @Test
    public void testStarredEventBasic() throws JSONException {
        CustomEvent event = MediaEventTemplate.newStarredTemplate().createEvent();

        EventTestUtils.validateEventValue(event, "event_name", MediaEventTemplate.STARRED_CONTENT_EVENT);
        EventTestUtils.validateEventValue(event, "template_type", "media");
        EventTestUtils.validateNestedEventValue(event, "properties", "ltv", "false");
    }

    /**
     * Test starred content event with optional properties.
     *
     * @throws JSONException
     */
    @Test
    public void testStarredEvent() throws JSONException {
        CustomEvent event = MediaEventTemplate.newStarredTemplate()
                                              .setCategory("media-category")
                                              .setId("starred-content-ID 1")
                                              .setDescription("This is a starred content media event.")
                                              .setType("audio type")
                                              .setAuthor("The Cool UA")
                                              .setFeature(true)
                                              .setPublishedDate("November 4, 2015")
                                              .createEvent();

        EventTestUtils.validateEventValue(event, "event_name", MediaEventTemplate.STARRED_CONTENT_EVENT);
        EventTestUtils.validateEventValue(event, "template_type", "media");
        EventTestUtils.validateNestedEventValue(event, "properties", "ltv", "false");
        EventTestUtils.validateNestedEventValue(event, "properties", "category", "\"media-category\"");
        EventTestUtils.validateNestedEventValue(event, "properties", "id", "\"starred-content-ID 1\"");
        EventTestUtils.validateNestedEventValue(event, "properties", "description", "\"This is a starred content media event.\"");
        EventTestUtils.validateNestedEventValue(event, "properties", "type", "\"audio type\"");
        EventTestUtils.validateNestedEventValue(event, "properties", "author", "\"The Cool UA\"");
        EventTestUtils.validateNestedEventValue(event, "properties", "feature", "true");
        EventTestUtils.validateNestedEventValue(event, "properties", "published_date", "\"November 4, 2015\"");
    }

    /**
     * Test shared content event.
     *
     * @throws JSONException
     */
    @Test
    public void testSharedEventBasic() throws JSONException {
        CustomEvent event = MediaEventTemplate.newSharedTemplate().createEvent();

        EventTestUtils.validateEventValue(event, "event_name", MediaEventTemplate.SHARED_CONTENT_EVENT);
        EventTestUtils.validateEventValue(event, "template_type", "media");
        EventTestUtils.validateNestedEventValue(event, "properties", "ltv", "false");
    }

    /**
     * Test shared content event with optional properties.
     *
     * @throws JSONException
     */
    @Test
    public void testSharedEvent() throws JSONException {
        CustomEvent event = MediaEventTemplate.newSharedTemplate("facebook", "social")
                                              .setCategory("media-category")
                                              .setId("shared-content-ID 2")
                                              .setDescription("This is a shared content media event.")
                                              .setType("video type")
                                              .setAuthor("The Cool UA")
                                              .setFeature(true)
                                              .setPublishedDate("November 4, 2015")
                                              .createEvent();

        EventTestUtils.validateEventValue(event, "event_name", MediaEventTemplate.SHARED_CONTENT_EVENT);
        EventTestUtils.validateEventValue(event, "template_type", "media");
        EventTestUtils.validateNestedEventValue(event, "properties", "ltv", "false");
        EventTestUtils.validateNestedEventValue(event, "properties", "source", "\"facebook\"");
        EventTestUtils.validateNestedEventValue(event, "properties", "medium", "\"social\"");
        EventTestUtils.validateNestedEventValue(event, "properties", "category", "\"media-category\"");
        EventTestUtils.validateNestedEventValue(event, "properties", "id", "\"shared-content-ID 2\"");
        EventTestUtils.validateNestedEventValue(event, "properties", "description", "\"This is a shared content media event.\"");
        EventTestUtils.validateNestedEventValue(event, "properties", "type", "\"video type\"");
        EventTestUtils.validateNestedEventValue(event, "properties", "author", "\"The Cool UA\"");
        EventTestUtils.validateNestedEventValue(event, "properties", "feature", "true");
        EventTestUtils.validateNestedEventValue(event, "properties", "published_date", "\"November 4, 2015\"");
    }

    /**
     * Test consumed content event.
     *
     * @throws JSONException
     */
    @Test
    public void testConsumedEventBasic() throws JSONException {
        CustomEvent event = MediaEventTemplate.newConsumedTemplate().createEvent();

        EventTestUtils.validateEventValue(event, "event_name", MediaEventTemplate.CONSUMED_CONTENT_EVENT);
        EventTestUtils.validateEventValue(event, "template_type", "media");
        EventTestUtils.validateNestedEventValue(event, "properties", "ltv", "false");
    }

    /**
     * Test consumed content event with optional properties.
     *
     * @throws JSONException
     */
    @Test
    public void testConsumedEvent() throws JSONException {
        CustomEvent event = MediaEventTemplate.newConsumedTemplate(2.99)
                                              .setCategory("media-category")
                                              .setId("consumed-content-ID 1")
                                              .setDescription("This is a consumed content media event.")
                                              .setType("audio type")
                                              .setAuthor("The Cool UA")
                                              .setFeature(true)
                                              .setPublishedDate("November 4, 2015")
                                              .createEvent();

        EventTestUtils.validateEventValue(event, "event_name", MediaEventTemplate.CONSUMED_CONTENT_EVENT);
        EventTestUtils.validateEventValue(event, "template_type", "media");
        EventTestUtils.validateEventValue(event, "event_value", 2990000.0);
        EventTestUtils.validateNestedEventValue(event, "properties", "ltv", "true");
        EventTestUtils.validateNestedEventValue(event, "properties", "category", "\"media-category\"");
        EventTestUtils.validateNestedEventValue(event, "properties", "id", "\"consumed-content-ID 1\"");
        EventTestUtils.validateNestedEventValue(event, "properties", "description", "\"This is a consumed content media event.\"");
        EventTestUtils.validateNestedEventValue(event, "properties", "type", "\"audio type\"");
        EventTestUtils.validateNestedEventValue(event, "properties", "author", "\"The Cool UA\"");
        EventTestUtils.validateNestedEventValue(event, "properties", "feature", "true");
        EventTestUtils.validateNestedEventValue(event, "properties", "published_date", "\"November 4, 2015\"");
    }
}

File: urbanairship-core/src/test/java/com/urbanairship/analytics/template/MediaEventTemplateTest.kt
/* Copyright Airship and Contributors */

package com.urbanairship.analytics.template

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.urbanairship.analytics.CustomEvent
import com.urbanairship.analytics.templates.MediaEventTemplate
import com.urbanairship.json.jsonMapOf
import junit.framework.TestCase.assertEquals
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
public class MediaEventTemplateTest {

    @Test
    public fun testBrowsed() {
        val event = CustomEvent.newBuilder(MediaEventTemplate.Type.Browsed).build()

        assertEquals("browsed_content", event.eventName)
        assertEquals("media", event.templateType)
        assertEquals(jsonMapOf("ltv" to false), event.properties)
    }

    @Test
    public fun testConsumed() {
        val event = CustomEvent.newBuilder(MediaEventTemplate.Type.Consumed).build()

        assertEquals("consumed_content", event.eventName)
        assertEquals("media", event.templateType)
        assertEquals(jsonMapOf("ltv" to false), event.properties)
    }

    @Test
    public fun testShared() {
        val event = CustomEvent.newBuilder(MediaEventTemplate.Type.Shared(
            source = "some source",
            medium = "some medium"
        )).build()

        assertEquals("shared_content", event.eventName)
        assertEquals("media", event.templateType)
        assertEquals(jsonMapOf(
            "ltv" to false,
            "source" to "some source",
            "medium" to "some medium"
        ), event.properties)
    }

    @Test
    public fun testSharedEmptyDetails() {
        val event = CustomEvent.newBuilder(MediaEventTemplate.Type.Shared()).build()

        assertEquals("shared_content", event.eventName)
        assertEquals("media", event.templateType)
        assertEquals(jsonMapOf(
            "ltv" to false,
        ), event.properties)
    }

    @Test
    public fun testStarred() {
        val event = CustomEvent.newBuilder(MediaEventTemplate.Type.Starred).build()

        assertEquals("starred_content", event.eventName)
        assertEquals("media", event.templateType)
        assertEquals(jsonMapOf("ltv" to false), event.properties)
    }

    @Test
    public fun testProperties() {
        val event = CustomEvent.newBuilder(
            type = MediaEventTemplate.Type.Shared(
                source = "some source",
                medium = "some medium"
            ),
            properties = MediaEventTemplate.Properties(
                id = "some id",
                category = "some category",
                type = "some type",
                eventDescription = "some description",
                isLTV = true,
                author = "some author",
                publishedDate = "date-as-string",
                isFeature = true
            )
        ).build()

        assertEquals("shared_content", event.eventName)
        assertEquals("media", event.templateType)
        assertEquals(jsonMapOf(
            "id" to "some id",
            "category" to "some category",
            "type" to "some type",
            "description" to "some description",
            "ltv" to true,
            "author" to "some author",
            "published_date" to "date-as-string",
            "feature" to true,
            "source" to "some source",
            "medium" to "some medium"
        ), event.properties)
    }
}
